executor/builder.go

```diff
@@ -1236,15 +1236,79 @@ func (b *executorBuilder) buildUpdate(v *plan.Update) Executor {
 		b.err = errors.Trace(b.err)
 		return nil
 	}
+	columns2Handle := buildColumns2Handle(v.Schema(), tblID2table)
 	updateExec := &UpdateExec{
-		baseExecutor: newBaseExecutor(b.ctx, nil, v.ExplainID(), selExec),
-		SelectExec:   selExec,
-		OrderedList:  v.OrderedList,
-		tblID2table:  tblID2table,
+		baseExecutor:   newBaseExecutor(b.ctx, nil, v.ExplainID(), selExec),
+		SelectExec:     selExec,
+		OrderedList:    v.OrderedList,
+		tblID2table:    tblID2table,
+		columns2Handle: columns2Handle,
 	}
 	return updateExec
 }
 
+// Columns2HandleEntry represents an mapper from column index to handle index.
+type Columns2HandleEntry struct {
+	start, end int
+	handleIdx  int
+}
+
+// Columns2Handle attaches the methods of sort.Interface to []Columns2HandleEntry sorting in increasing order.
+type Columns2Handle []Columns2HandleEntry
+
+// Len implements sort.Interface#Len.
+func (c Columns2Handle) Len() int {
+	return len(c)
+}
+
+// Len implements sort.Interface#Swap.
+func (c Columns2Handle) Swap(i, j int) {
+	c[i], c[j] = c[j], c[i]
+}
+
+// Len implements sort.Interface#Less.
+// let ranges first sorted by `start` increasing order, and then sorted by `end` increasing order if `start` are equal.
+func (c Columns2Handle) Less(i, j int) bool {
+	if c[i].start == c[j].start {
+		return c[i].end < c[i].end
+	}
+	return c[i].start < c[j].start
+}
+
+// findHandle finds the range hit by given column index.
+// c must be sorted using sort.Sort.
+func (c Columns2Handle) findHandle(colIndex int) (int, bool) {
+	if c == nil || len(c) == 0 {
+		return 0, false
+	}
+	biggerOne := sort.Search(len(c), func(i int) bool { return c[i].start > colIndex })
+	if biggerOne == 0 {
+		return 0, false
+	}
+	if c[biggerOne-1].start <= colIndex && colIndex < c[biggerOne-1].end {
+		return c[biggerOne-1].handleIdx, true
+	}
+	return 0, false
+}
+
+// buildColumns2Handle build columns to handle mapping.
+func buildColumns2Handle(schema *expression.Schema, tblID2Table map[int64]table.Table) Columns2Handle {
+	if len(schema.TblID2Handle) < 2 {
+		return nil
+	}
+	var cols2Handle Columns2Handle
+	for tblID, handleCols := range schema.TblID2Handle {
+		tbl := tblID2Table[tblID]
+		for _, handleCol := range handleCols {
+			offset := getTableOffset(schema, handleCol)
+			end := offset + len(tbl.WritableCols())
+			cols2Handle = append(cols2Handle, Columns2HandleEntry{offset, end, handleCol.Index})
+		}
+	}
+	sort.Sort(cols2Handle)
```

Comment:
Do we need to sort them?

Reply:
for later find at https://github.com/pingcap/tidb/pull/7177/files/d67a05e0fcf141855ede1651256e10a85f41cf0f#diff-badfcd30d7596a08cd207b8e6ae778e6R1284 we need sorted, and for-loop TblID2Handle doesn't make sure order o o?